Initial support for tags

zquestz opened the pull request:
I have added Tags() to the provider interface so you can easily launch multiple searches at the same time. Figured it is useful when trying to search for movies, news, etc.

- [x] Updated all auto complete scripts to include the new flags.
- [x] Supports searching a tag and a specific provider. Useful if you want to search a tag, but also google.
- [x] Supports adding tags to custom providers. Just add "tags" to your config.
- [x] Added `--list-tags` flag to display all available tags.
- [x] Added Tags() method to all providers.
- [x] Updated README.md
- [x] Added support for tag expansion.
- [x] Populate all providers with an initial set of tags.
- [x] Update -v flag to support list views.

akb commented:
Idea: environment conditional tags (e.g. only show arch stuff if using arch)

"photos": 500px, 
"streaming": 8tracks bandcamp
"music": 8tracks bandcamp
"shopping": amazon
"software": archpkg atmospherejs aur codepen
"arch": archpkg archwiki aur
"info": arxiv
"papers": arxiv
"javascript": atmospherejs codepen
"general": baidu bing
"news": arstechnica bgr buzzfeed cnn
"tech": arstechnica bgr
"bullshit": buzzfeed

akb commented:
"arch": archpkg archwiki aur
"code": dockerhub github stackoverflow
"education": arxiv coursera libgen wikipedia
"forums": quora reddit
"general": bing duckduckgo google yahoo
"go": go godoc
"javascript": atmospherejs npm
"movies": crunchyroll netflix
"music": 8tracks bandcamp soundcloud spotify
"news": cnn npr theregister
"photos": 500px flickr instagram
"pics": dribbble flickr imgur instagram
"shopping": amazon overstock
"social": facebook googleplus twitter
"tech-news": arstechnica bgr engadget hackernews
"torrent": kickasstorrents thepiratebay torrentz
"video": vimeo youtube

zquestz commented:
Yeah that looks like a reasonable start. Thanks @akb!

shr3kst3r commented:
Maybe change "general" to "search".

zquestz commented:
@shr3kst3r I was thinking the same thing. =)

zquestz commented:
Probably will change torrent to torrents as well.

zquestz commented:
What do you guys think of ecom instead of shopoing?

zquestz commented:
Ok this is now ready for review. I went ahead and added tag expansion. This will search the tag tech-news for ssd info.

```
s -t te ssd
```

I also added the tags as suggested, but changed general to search, and removed dockerhub from code, as I didn't think it fit.

zquestz commented:
Also went ahead and fixed the list views. Here is sample output:

s -l -v
```
500px (photos)
8tracks (music)
amazon (shopping)
archpkg (arch)
archwiki (arch)
arstechnica (tech-news)
arxiv (education)
atmospherejs (javascript)
aur (arch)
baidu
bandcamp (music)
bgr (tech-news)
bing (search)
buzzfeed
cnn (news)
codepen
coursera (education)
cplusplus
crunchyroll (movies)
debianpkg
digg
diigo
dockerhub
dribbble (pics)
duckduckgo (search)
dumpert
engadget (tech-news)
facebook (social)
flickr (photos, pics)
flipkart
foursquare
gist
github (code)
gmail
go (go)
godoc (go)
google (search)
googledocs
googleplus (social)
hackernews (tech-news)
ietf
ifttt
imdb (movies)
imgur (pics)
inbox
instagram (photos, pics)
kickasstorrents (torrents)
libgen (education)
linkedin
lmgtfy
macports
mdn
metacpan
msdn
naver
netflix (movies)
nhaccuatui
npm (javascript)
npmsearch (javascript)
npr (news)
nvd
overstock (shopping)
packagist
phandroid
php
pinterest
python
quora (forums)
reddit (forums)
rottentomatoes
rubygems
soundcloud (music)
spotify (music)
stackoverflow (code)
steam
taobao
thepiratebay (torrents)
theregister (news)
torrentz (torrents)
twitchtv
twitter (social)
unity3d
upcloud
vimeo (video)
wikipedia (education)
wolframalpha
yahoo (search)
yandex
youtube (video)
```

tw4dl commented:
rottentomatoes (movies)
gist (code)

zquestz commented:
@tw4dl thanks, those were adjusted. =)

lilirui commented on providers/taobao/taobao.go:
```diff
@@ -18,3 +18,8 @@ type Provider struct{}
+
+// Tags returns the tags relevant to this provider.
+func (p *Provider) Tags() []string {
+	return []string{}
```
This should be 'shopping' also.

zquestz replied:
Good call, one sec.

lilirui commented:
taobao (shopping)
baidu (search)

zquestz commented:
Added =)

zquestz commented:
@KeizerDev any thoughts?

zquestz commented:
Going to merge this in about an hour. If anyone has suggestions or changes please let me know!

tw4dl commented:
yandex (search)
inbox gmail (email)

zquestz commented:
Not going to do an email tag as most users just use one main mail provider. Yandex was added. =)
